File: src/main/java/miragefairy2019/colormaker/WindowColorMaker.java
package miragefairy2019.colormaker;

import kotlin.Unit;
import miragefairy2019.colormaker.core.ColorConstants;
import miragefairy2019.colormaker.core.ImageLayer;
import miragefairy2019.colormaker.core.LabelImage;
import miragefairy2019.colormaker.core.PanelColorSlider;
import mirrg.boron.util.struct.ImmutableArray;

import javax.swing.JFrame;
import javax.swing.JPanel;
import javax.swing.JTextField;
import javax.swing.border.EmptyBorder;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.GridLayout;
import java.awt.Insets;
import java.util.stream.Stream;

import static mirrg.boron.swing.UtilsComponent.createPanelTitledBorder;
import static mirrg.boron.swing.UtilsComponent.createSplitPaneHorizontal;
import static mirrg.boron.swing.UtilsComponent.get;

public class WindowColorMaker extends JFrame {

    private JPanel panelImage;
    private LabelImage labelMirageFairy;
    private LabelImage labelMirageWisp;
    private LabelImage labelMagicSphere;
    private PanelColorSlider panelColorSliderBG;
    private PanelColorSlider panelColorSlider0;
    private PanelColorSlider panelColorSlider1;
    private PanelColorSlider panelColorSlider2;
    private PanelColorSlider panelColorSlider3;
    private JTextField textFieldColors;

    private boolean isInProcessing = false;
    private boolean leadyUpdate = false;

    public WindowColorMaker() {

        getContentPane().setLayout(new GridLayout(0, 1, 0, 0));

        add(createSplitPaneHorizontal(

            // 左ペイン
            get(panelImage = new JPanel(), c2 -> {
                c2.setLayout(get(new GridBagLayout(), l -> {
                    l.columnWidths = new int[]{100};
                    l.rowHeights = new int[]{0, 0, 0};
                    l.columnWeights = new double[]{0.0};
                    l.rowWeights = new double[]{0.0, 0.0, 0.0};
                }));

                // 画像3枚
                {
                    ColorConstants colorConstants = new ColorConstants();
                    colorConstants.addConstant("@skin", () -> value[0]);
                    colorConstants.addConstant("@darker", () -> value[1]);
                    colorConstants.addConstant("@brighter", () -> value[2]);
                    colorConstants.addConstant("@hair", () -> value[3]);

                    c2.add(get(labelMirageFairy = new LabelImage(), c -> {
                        c.setPreferredSize(new Dimension(64, 64));
                        c.setColorConstants(colorConstants);
                    }), get(new GridBagConstraints(), c -> {
                        c.insets = new Insets(0, 0, 5, 0);
                        c.gridx = 0;
                        c.gridy = 0;
                    }));
                    c2.add(get(labelMirageWisp = new LabelImage(), c -> {
                        c.setPreferredSize(new Dimension(64, 64));
                        c.setColorConstants(colorConstants);
                    }), get(new GridBagConstraints(), c -> {
                        c.insets = new Insets(0, 0, 5, 0);
                        c.gridx = 0;
                        c.gridy = 1;
                    }));
                    c2.add(get(labelMagicSphere = new LabelImage(), c -> {
                        c.setPreferredSize(new Dimension(64, 64));
                        c.setColorConstants(colorConstants);
                    }), get(new GridBagConstraints(), c -> {
                        c.gridx = 0;
                        c.gridy = 2;
                    }));

                }

            }),

            // 右ペイン
            get(new JPanel(), c2 -> {
                c2.setBorder(new EmptyBorder(4, 4, 4, 4));
                c2.setLayout(get(new GridBagLayout(), l -> {
                    l.columnWidths = new int[]{0};
                    l.rowHeights = new int[]{0, 0, 0, 0, 0, 0};
                    l.columnWeights = new double[]{1.0};
                    l.rowWeights = new double[]{0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
                }));

                // 右側スライダーコンポーネント
                c2.add(createPanelTitledBorder("Background", get(
                    panelColorSliderBG = new PanelColorSlider(), c -> {
                        c.getListeners().add(color -> {
                            if (isInProcessing) return Unit.INSTANCE;
                            setBackgroundColor(color, c);
                            return Unit.INSTANCE;
                        });
                    })), get(new GridBagConstraints(), c -> {
                    c.fill = GridBagConstraints.HORIZONTAL;
                    c.insets = new Insets(0, 0, 5, 0);
                    c.gridx = 0;
                    c.gridy = 0;
                }));
                c2.add(createPanelTitledBorder("Skin", get(
                    panelColorSlider0 = new PanelColorSlider(), c -> {
                        c.getListeners().add(color -> {
                            if (isInProcessing) return Unit.INSTANCE;
                            setValue(getColor(), c);
                            return Unit.INSTANCE;
                        });
                    })), get(new GridBagConstraints(), c -> {
                    c.fill = GridBagConstraints.HORIZONTAL;
                    c.insets = new Insets(0, 0, 5, 0);
                    c.gridx = 0;
                    c.gridy = 1;
                }));
                c2.add(createPanelTitledBorder("Darker", get(
                    panelColorSlider1 = new PanelColorSlider(), c -> {
                        c.getListeners().add(color -> {
                            if (isInProcessing) return Unit.INSTANCE;
                            setValue(getColor(), c);
                            return Unit.INSTANCE;
                        });
                    })), get(new GridBagConstraints(), c -> {
                    c.fill = GridBagConstraints.HORIZONTAL;
                    c.insets = new Insets(0, 0, 5, 0);
                    c.gridx = 0;
                    c.gridy = 2;
                }));
                c2.add(createPanelTitledBorder("Brighter", get(
                    panelColorSlider2 = new PanelColorSlider(), c -> {
                        c.getListeners().add(color -> {
                            if (isInProcessing) return Unit.INSTANCE;
                            setValue(getColor(), c);
                            return Unit.INSTANCE;
                        });
                    })), get(new GridBagConstraints(), c -> {
                    c.fill = GridBagConstraints.HORIZONTAL;
                    c.insets = new Insets(0, 0, 5, 0);
                    c.gridx = 0;
                    c.gridy = 3;
                }));
                c2.add(createPanelTitledBorder("Hair", get(
                    panelColorSlider3 = new PanelColorSlider(), c -> {
                        c.getListeners().add(color -> {
                            if (isInProcessing) return Unit.INSTANCE;
                            setValue(getColor(), c);
                            return Unit.INSTANCE;
                        });
                    })), get(new GridBagConstraints(), c -> {
                    c.insets = new Insets(0, 0, 5, 0);
                    c.fill = GridBagConstraints.HORIZONTAL;
                    c.gridx = 0;
                    c.gridy = 4;
                }));

                // 右側色構文
                c2.add(get(textFieldColors = new JTextField(), c -> {
                    c.setColumns(10);
                    c.addActionListener(e -> {
                        if (isInProcessing) return;
                        setValue(c.getText(), c);
                    });
                    c.getDocument().addDocumentListener(new DocumentListener() {
                        @Override
                        public void removeUpdate(DocumentEvent e) {
                            if (isInProcessing) return;
                            setValue(c.getText(), c);
                        }

                        @Override
                        public void insertUpdate(DocumentEvent e) {
                            if (isInProcessing) return;
                            setValue(c.getText(), c);
                        }

                        @Override
                        public void changedUpdate(DocumentEvent e) {
                            if (isInProcessing) return;
                            setValue(c.getText(), c);
                        }
                    });
                }), get(new GridBagConstraints(), c -> {
                    c.fill = GridBagConstraints.HORIZONTAL;
                    c.gridx = 0;
                    c.gridy = 5;
                }));

            })

        ));

        //

        setBackgroundColor(new Color(139, 139, 139));
        setValue(new Color[]{
            Color.white,
            Color.white,
            Color.white,
            Color.white,
        });

        leadyUpdate = true;
        updateImage();

        setLocationByPlatform(true);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        pack();
    }

    private Color[] getColor() {
        return new Color[]{
            panelColorSlider0.getValue(),
            panelColorSlider1.getValue(),
            panelColorSlider2.getValue(),
            panelColorSlider3.getValue(),
        };
    }

    //

    private ImmutableArray<ImmutableArray<ImageLayer>> images = MainColorMaker.loadImages();

    public void updateImage() {
        if (!leadyUpdate) return;
        labelMirageFairy.setImage(images.get(0));
        labelMirageWisp.setImage(images.get(1));
        labelMagicSphere.setImage(images.get(2));
    }

    //

    private Color[] value;

    private void setValue(String text, Object source) {
        Color[] colors;
        try {
            colors = Stream.of(text.split(","))
                .map(s -> Color.decode(s.trim()))
                .toArray(Color[]::new);
        } catch (RuntimeException e) {
            return;
        }
        if (colors.length != 4) return;

        setValue(colors, source);
    }

    private void setValue(Color[] value) {
        setValue(value, null);
    }

    private void setValue(Color[] value, Object source) {
        isInProcessing = true;

        this.value = value;
        if (source != panelColorSlider0) panelColorSlider0.setValue(value[0]);
        if (source != panelColorSlider1) panelColorSlider1.setValue(value[1]);
        if (source != panelColorSlider2) panelColorSlider2.setValue(value[2]);
        if (source != panelColorSlider3) panelColorSlider3.setValue(value[3]);
        if (source != textFieldColors) textFieldColors.setText(String.format("0x%06X, 0x%06X, 0x%06X, 0x%06X",
            value[0].getRGB() & 0xffffff,
            value[1].getRGB() & 0xffffff,
            value[2].getRGB() & 0xffffff,
            value[3].getRGB() & 0xffffff));
        updateImage();

        isInProcessing = false;
    }

    //

    private void setBackgroundColor(Color backgroundColor) {
        setBackgroundColor(backgroundColor, null);
    }

    private void setBackgroundColor(Color backgroundColor, Object source) {
        isInProcessing = true;

        panelImage.setBackground(backgroundColor);
        labelMirageFairy.setBackgroundColor(backgroundColor);
        labelMirageWisp.setBackgroundColor(backgroundColor);
        labelMagicSphere.setBackgroundColor(backgroundColor);
        if (source != panelColorSliderBG) panelColorSliderBG.setValue(backgroundColor);
        updateImage();

        isInProcessing = false;
    }

}

File: src/main/java/miragefairy2019/colormaker/WindowColorMaker.kt
package miragefairy2019.colormaker

import mirrg.boron.swing.UtilsComponent
import mirrg.boron.swing.UtilsComponent.createPanelTitledBorder
import mirrg.kotlin.hydrogen.formatAs
import mirrg.kotlin.hydrogen.join
import java.awt.Color
import java.awt.Dimension
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.GridLayout
import java.awt.Insets
import java.awt.image.BufferedImage
import javax.swing.JFrame
import javax.swing.JPanel
import javax.swing.JTextField
import javax.swing.border.EmptyBorder
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener

private fun String.decodeColor() = try {
    Color.decode(this)
} catch (_: NumberFormatException) {
    null
}

private fun String.decodeColorsOrNull(): List<Color>? = this.split(",").map { it.trim().decodeColor() ?: return null }

private fun List<Color>.encode() = this.map { it.rgb and 0xffffff formatAs "0x%06X" }.join(", ")

private class ExclusiveController {
    private var isInProcessing = false
    operator fun invoke(block: () -> Unit) {
        if (isInProcessing) return
        isInProcessing = true
        try {
            block()
        } finally {
            isInProcessing = false
        }
    }
}


class LayeredImageSetting(val layerSettings: List<LayerSetting>)

fun layeredImageSettingOf(vararg layerSettings: LayerSetting) = LayeredImageSetting(layerSettings.toList())

class LayerSetting(val imagePath: String, val colorExpression: ColorExpression)

class WindowColorMaker(
    private val imageLoader: (String) -> BufferedImage,
    private val layeredImageSettings: List<LayeredImageSetting>,
    private val sliderNames: List<String>
) : JFrame() {

    // Utility
    private val exclusiveController = ExclusiveController()

    // 画像バッファ
    private val imageLayerListList = layeredImageSettings.map { layeredImageSetting ->
        layeredImageSetting.layerSettings.map { layerSetting ->
            Layer(imageLoader(layerSetting.imagePath), layerSetting.colorExpression)
        }
    }

    // Components
    private val panelImage: JPanel
    private val labelImages = mutableListOf<LayeredImage>()
    private val panelColorSliderBG: PanelColorSlider
    private val panelColorSliders = mutableListOf<PanelColorSlider>()
    private val textFieldColors: JTextField


    // Value

    private fun setBackgroundColor(backgroundColor: Color, source: Any?) {
        exclusiveController {
            if (source !== panelImage) panelImage.background = backgroundColor
            labelImages.forEach { labelImage ->
                if (source !== labelImage) labelImage.backgroundColor = backgroundColor
            }
            if (source !== panelColorSliderBG) panelColorSliderBG.setValue(backgroundColor)
            updateImage()
        }
    }

    private fun getColor(index: Int) = panelColorSliders[index].getValue()

    private fun setColors(colors: List<Color>, source: Any?) {
        exclusiveController {
            panelColorSliders.indices.forEach { i ->
                if (source !== panelColorSliders[i]) panelColorSliders[i].setValue(colors[i])
            }
            if (source !== textFieldColors) textFieldColors.text = colors.encode()
            updateImage()
        }
    }

    private var canUpdateImage = false

    private fun updateImage() {
        if (!canUpdateImage) return
        labelImages.forEachIndexed { i, it ->
            it.render(imageLayerListList[i])
        }
    }


    init {

        contentPane.layout = GridLayout(0, 1, 0, 0)

        add(UtilsComponent.createSplitPaneHorizontal(

            // 左ペイン
            JPanel().also { leftPane ->
                panelImage = leftPane

                leftPane.layout = GridBagLayout().also {
                    it.columnWidths = intArrayOf(100)
                    it.rowHeights = intArrayOf(0, 0, 0)
                    it.columnWeights = doubleArrayOf(0.0)
                    it.rowWeights = doubleArrayOf(0.0, 0.0, 0.0)
                }

                // 画像
                val colorEvaluator = ColorEvaluator().also {
                    sliderNames.forEachIndexed { i, name ->
                        it.registerVariable("@$name") { getColor(i) }
                    }
                }
                repeat(layeredImageSettings.size) { i ->
                    leftPane.add(LayeredImage().also { labelImage ->
                        labelImages += labelImage
                        labelImage.preferredSize = Dimension(64, 64)
                        labelImage.colorEvaluator = colorEvaluator
                    }, GridBagConstraints().also {
                        it.insets = Insets(0, 0, 5, 0)
                        it.gridx = 0
                        it.gridy = i
                    })
                }

            },

            // 右ペイン
            JPanel().also { rightPane ->

                rightPane.border = EmptyBorder(4, 4, 4, 4)
                rightPane.layout = GridBagLayout().also {
                    it.columnWidths = intArrayOf(0)
                    it.rowHeights = intArrayOf(0, 0, 0, 0, 0, 0)
                    it.columnWeights = doubleArrayOf(1.0)
                    it.rowWeights = doubleArrayOf(0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
                }

                // 右側スライダーコンポーネント
                rightPane.add(createPanelTitledBorder("Background",
                    PanelColorSlider().also { c ->
                        panelColorSliderBG = c
                        c.listeners += { color -> setBackgroundColor(color, c) }
                    }), GridBagConstraints().also {
                    it.fill = GridBagConstraints.HORIZONTAL
                    it.insets = Insets(0, 0, 5, 0)
                    it.gridx = 0
                    it.gridy = 0
                })
                sliderNames.forEachIndexed { i, name ->
                    rightPane.add(createPanelTitledBorder(name,
                        PanelColorSlider().also { c ->
                            panelColorSliders += c
                            c.listeners += { setColors(panelColorSliders.map { it.getValue() }, c) }
                        }), GridBagConstraints().also {
                        it.fill = GridBagConstraints.HORIZONTAL
                        it.insets = Insets(0, 0, 5, 0)
                        it.gridx = 0
                        it.gridy = 1 + i
                    })
                }

                // 右側色構文
                rightPane.add(JTextField().also { c ->
                    textFieldColors = c
                    c.columns = 10
                    fun setSliderValue() {
                        val colors = c.text.decodeColorsOrNull() ?: return
                        if (colors.size != sliderNames.size) return
                        setColors(colors, c)
                    }
                    c.addActionListener { setSliderValue() }
                    c.document.addDocumentListener(object : DocumentListener {
                        override fun removeUpdate(e: DocumentEvent) = setSliderValue()
                        override fun insertUpdate(e: DocumentEvent) = setSliderValue()
                        override fun changedUpdate(e: DocumentEvent) = setSliderValue()
                    })
                }, GridBagConstraints().also {
                    it.fill = GridBagConstraints.HORIZONTAL
                    it.gridx = 0
                    it.gridy = sliderNames.size + 1
                })

            }

        ))


        setBackgroundColor(Color(139, 139, 139), null)
        setColors(sliderNames.map { Color.white!! }, null)

        canUpdateImage = true
        updateImage()

        isLocationByPlatform = true
        defaultCloseOperation = DISPOSE_ON_CLOSE
        pack()
    }
}
